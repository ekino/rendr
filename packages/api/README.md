---
<p align="center">
<strong>
The package is part of the `@ekino/rendr` project. Please visit <a aria-label=rendr learn" href="https://github.com/ekino/rendr">https://github.com/ekino/rendr</a> to get started with rendr.
</strong>
</p>
---

## Contributing

Please see our [CONTRIBUTING.md](https://github.com/ekino/rendr/CONTRIBUTING.md).

## Authors

- Thomas Rabaix ([@th0masr](https://twitter.com/th0masr)) â€“ [Ekino](https://ekino.com)
