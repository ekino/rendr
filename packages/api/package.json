{
    "name": "@ekino/rendr-api",
    "license": "MIT",
    "version": "0.0.1-DEV",
    "main": "dist/lib/index.js",
    "typing": "dist/lib/index.d.ts",
    "files": [
        "lib"
    ],
    "publishConfig": {
        "access": "public"
    },
    "dependencies": {
        "@ekino/rendr-core": "0.0.1-DEV",
        "@ekino/rendr-loader": "0.0.1-DEV",
        "express": "^4.16.4"
    },
    "devDependencies": {
        "@types/express": "^4.16.1",
        "@types/node": "^12.6.9"
    },
    "scripts": {
        "build": "tsc"
    }
}
