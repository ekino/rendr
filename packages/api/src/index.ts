import { Loader } from "@ekino/rendr-loader";

import Express from "express";

// should return a middleware api to return page definition through http
export function createApi(loader: Loader): Express.RequestHandler {
  return async (req, res, next) => {
    const ctx = {
      req,
      res,
      pathname: "",
      params: {},
      query: "",
      asPath: req.url
    };

    res.set("X-Rendr-Content-Type", "rendr/octet-stream");

    const page = await loader(ctx);

    if (!page) {
      return;
    }

    res.set("X-Rendr-Content-Type", "rendr/document");

    res.json(page);
  };
}
