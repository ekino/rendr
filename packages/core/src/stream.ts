import { Readable, Writable } from "stream";
import { Page, createPage } from "./index";
import { Body } from "./types";

/**
 * This function pipe a Readable Stream into a Writable stream in order to pass
 * the response to the server from to the client. This is usefull to act as a proxy
 * and avoid to buffer all the response body in the current process.
 *
 * @param source
 * @param dest
 */
export function pipe(source: Body, dest: Writable): Promise<void> {
  if (typeof source == "string") {
    source = Readable.from(source);
  }

  return new Promise<void>((resolve, reject) => {
    if (!(source instanceof Readable)) {
      return reject("Not a Readable object");
    }

    source.pipe(dest);

    source.on("end", (err: Error) => {
      if (err) {
        reject(err);
      } else {
        resolve();
      }
    });
  });
}

/**
 * This function pipe a Readable Stream into an internal buffer so it is possible
 * to reuse the buffer to create a Page object.
 *
 * @param source
 */
export function pipePageToClient(source: Body): Promise<Page> {
  if (typeof source == "string") {
    source = Readable.from(source);
  }

  let data = "";
  const dest = new Writable({
    write(chunk, _, callback) {
      data += chunk;

      callback();
    },
  });

  return pipe(source, dest).then(() => {
    return createPage(JSON.parse(data));
  });
}
