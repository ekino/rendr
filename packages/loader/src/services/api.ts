import { RequestCtx, createPage, Page } from "@ekino/rendr-core";
import { Readable, Writable } from "stream";
import Axios from "axios";

import { Loader } from "../types";

const headersToTransfers = [
  "Last-Modified",
  "Content-Type",
  "Content-Range",
  "Content-Length",
  "Accept-Ranges",
  "Date"
];

// The loader is used by NextJS to load the Page definition from the API
// This code can be called from the nodejs or the browser.
// However the pipe for raw binary are not likely to be run by the browser
// as the server will stream the content, and so cannot be run be the browser.
export function createApiLoader(baseUrl: string): Loader {
  return async (ctx: RequestCtx) => {
    const url = `${baseUrl}${ctx.req.url}`;

    const headers: any = {
      "X-User-Agent": "ekino/rendr"
    };

    // res and req are only avaible on the node runtime.
    if (ctx.res && ctx.req) {
      if ("range" in ctx.req.headers) {
        headers["range"] = ctx.req.headers["range"];
      }

      headers["accept"] = ctx.req.headers["accept"];
      headers["accept-encoding"] = ctx.req.headers["accept-encoding"];
    }

    const response = await Axios.get(url, {
      responseType: "stream",
      headers
    });

    if (!("x-rendr-content-type" in response.headers)) {
      // @todo: check how we can add a logger here
      return;
    }

    if (response.headers["x-rendr-content-type"] === "rendr/document") {
      return pipeToPage(response.data);
    }

    // This code should be called only from node, there is reason why this code should be called on the client
    if (
      response.headers["x-rendr-content-type"] !== "rendr/document" &&
      ctx.res &&
      ctx.req
    ) {
      ctx.res.statusCode = response.status;

      headersToTransfers.forEach(n => {
        if (n.toLowerCase() in response.headers) {
          ctx.res.setHeader(n, response.headers[n.toLowerCase()]);
        }
      });

      return pipe(
        response.data,
        ctx.res
      );
    }
  };
}

/**
 * This function pipe a Readable Stream into a Writable stream in order to pass
 * the response to the server from to the client. This is usefull to act as a proxy
 * and avoid to buffer all the response body in the current process.
 *
 * @param source
 * @param dest
 */
function pipe(source: Readable, dest: Writable): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    source.pipe(dest);
    source.on("end", (err: Error) => {
      if (err) {
        reject(err);
      } else {
        resolve();
      }
    });
  });
}

/**
 * This function pipe a Readable Stream into an internal buffer so it is possible
 * to reuse the buffer to create a Page object.
 *
 * @param source
 */
function pipeToPage(source: Readable): Promise<Page> {
  let data = "";
  const dest = new Writable({
    write: chunk => {
      data += chunk;
    }
  });

  return pipe(
    source,
    dest
  ).then(() => {
    return createPage(JSON.parse(data));
  });
}
