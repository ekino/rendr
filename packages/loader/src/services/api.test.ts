import { createPage } from "@ekino/rendr-core";
import { IncomingMessage, ServerResponse } from "http";
import axios from "axios";
import { Readable, Writable } from "stream";

import { createApiLoader } from "./api";

jest.mock("axios");

describe("test API Loader", () => {
  it("test rendr/document mode, ie a Page", async () => {
    const loader = createApiLoader("/api");

    const req = jest.fn<IncomingMessage, any[]>();
    // @ts-ignore
    req.headers = {
      accept: "text/html",
      "accept-encoding": "gzip",
    };

    const rawData = JSON.stringify(createPage({}));
    let largeData = "[";
    for (let i = 0; i < 128; i++) {
      largeData += rawData + ",";
    }
    largeData += "{}]";

    let pos = 0;
    let cptCall = 0;
    const resp = {
      headers: {
        "x-rendr-content-type": "rendr/document",
      },
      data: new Readable({
        read(size) {
          const content = largeData.substr(pos, size);
          pos = pos + size;
          cptCall++;
          this.push(content.length ? content : null);
        },
      }),
    };
    // @ts-ignore
    axios.get.mockResolvedValue(resp);

    const res = jest.fn<ServerResponse, any[]>();

    const ctx = {
      pathname: "/",
      query: {},
      req: req,
      res: res,
      isServerSide: true,
      isClientSide: false,
    };

    // @ts-ignore
    const result = await loader(ctx);

    expect(result).toMatchSnapshot();
    expect(cptCall).toBe(3);
  });

  it("test non rendr/document mode, ie a Binary File", async () => {
    const loader = createApiLoader("/api");

    const req = jest.fn<IncomingMessage, any[]>();
    // @ts-ignore
    req.headers = {
      accept: "text/html",
      "accept-encoding": "gzip",
    };

    let strPage = "the content to be streamed...";
    let pos = 0;
    const resp = {
      headers: {
        "x-rendr-content-type": "rendr/octet-stream",
      },
      data: new Readable({
        read(size) {
          const content = strPage.substr(pos, size);
          pos = pos + size;

          this.push(content.length ? content : null);
        },
      }),
    };

    // @ts-ignore
    axios.get.mockResolvedValue(resp);

    let data = "";
    const res = new Writable({
      write(chunk) {
        data += chunk;
      },
    });

    const ctx = {
      pathname: "/",
      query: {},
      params: {},
      req: req,
      res: res,
      isServerSide: true,
      isClientSide: false,
    };

    // @ts-ignore
    const result = await loader(ctx);

    expect(data).toBe(strPage);
  });
});
