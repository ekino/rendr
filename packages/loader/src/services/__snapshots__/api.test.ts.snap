// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test API Loader test rendr/document mode, ie a Page 1`] = `
Page {
  "blocks": Array [],
  "cache": Object {
    "ttl": 0,
  },
  "head": Object {
    "defaultTitle": "-",
    "htmlAttributes": Object {},
    "link": "",
    "meta": Array [],
    "title": "-",
    "titleTemplate": "%s",
  },
  "id": "default",
  "settings": Array [],
  "statusCode": 200,
  "template": "default",
  "type": "document",
}
`;
